utils/update_checkout/update_checkout/update_checkout.py

```diff
@@ -171,9 +171,13 @@ def update_single_repository(pool_args):
                                                             checkout_target)
 
             # The clean option restores a repository to pristine condition.
-            if should_clean:
-                shell.run(['git', 'clean', '-fdx'],
-                          echo=True, prefix=prefix)
+            if should_clean or should_stash:
+                if should_stash:
+                    shell.run(['git', 'stash'],
+                              echo=True, prefix=prefix)
+                elif should_clean:
+                    shell.run(['git', 'clean', '-fdx'],
+                              echo=True, prefix=prefix)
                 shell.run(['git', 'submodule', 'foreach', '--recursive',
                            'git', 'clean', '-fdx'],
                           echo=True, prefix=prefix)
```

Comment:
I only just spotted this, sorry. I think there is a problem with the flow: 
* `--clean`:
  * Run `git clean` on the repository and its submodules (OK)
* `--stash`:
  * Run `git stash` on the repository (OK), but not submodules (?)
  * Run `git clean` on submodules (?)

Another thing I noticed is that the subsequent `git reset --hard` commands are redundant in the `--stash` case because stashing, unlike cleaning, resets the index and working directory.